src/utils/get-sorted-nodes-by-import-order.ts

```diff
@@ -0,0 +1,74 @@
+import { clone } from 'lodash';
+
+import { THIRD_PARTY_MODULES_SPECIAL_WORD, newLineNode } from '../constants';
+import { naturalSort } from '../natural-sort';
+import { GetSortedNodes, ImportGroups, ImportOrLine } from '../types';
+import { getImportNodesMatchedGroup } from './get-import-nodes-matched-group';
+import { getSortedImportSpecifiers } from './get-sorted-import-specifiers';
+
+/**
+ * This function returns the given nodes, sorted in the order as indicated by
+ * the importOrder array from the given options.
+ * The plugin considers these import nodes as local import declarations.
+ * @param nodes A subset of all import nodes that should be sorted.
+ * @param options Options to influence the behavior of the sorting algorithm.
+ */
+export const getSortedNodesByImportOrder: GetSortedNodes = (nodes, options) => {
+    naturalSort.insensitive = options.importOrderCaseInsensitive;
+
+    let { importOrder } = options;
+    const { importOrderSeparation, importOrderSortSpecifiers } = options;
+
+    const originalNodes = nodes.map(clone);
+    const finalNodes: ImportOrLine[] = [];
+
+    if (!importOrder.includes(THIRD_PARTY_MODULES_SPECIAL_WORD)) {
+        importOrder = [THIRD_PARTY_MODULES_SPECIAL_WORD, ...importOrder];
+    }
+
+    const importOrderGroups = importOrder.reduce<ImportGroups>(
+        (groups, regexp) => ({
+            ...groups,
+            [regexp]: [],
+        }),
+        {},
+    );
+
+    const importOrderWithOutThirdPartyPlaceholder = importOrder.filter(
+        (group) => group !== THIRD_PARTY_MODULES_SPECIAL_WORD,
+    );
+
+    for (const node of originalNodes) {
+        const matchedGroup = getImportNodesMatchedGroup(
+            node,
+            importOrderWithOutThirdPartyPlaceholder,
+        );
+        importOrderGroups[matchedGroup].push(node);
+    }
+
+    for (const group of importOrder) {
+        const groupNodes = importOrderGroups[group];
+        const last = group === importOrder[importOrder.length - 1];
```

Comment:
I changed this, no problem, but I'm curious, do you have bad experiences with these kind of loops?

Reply:
It wasn't about the loop, it was about the definition of `last`. I thought it was unnecessary to access `importOrder` twice each time to make a string comparison when a simple index comparison could do the job and it's more intuitive for the definition of `last`. The performance aspect of this suggestion was not tested but I thought it could play a role since we do not necessarily know how long `importOrder` is and how `group` looks like.

Reply:
Ah I see what you mean, sorry, I guess it must've been pretty late yesterday ; ) Considering that in each iteration of the loop we are sorting the imports of that group, which performs many string comparisons, that one comparison probably doesn't matter much.

But you're right, checking the index is better and easier to read anyways (it would fail if e.g. the first group has the same name as the last), so let me change that.